toolkit/tools/pkg/imagecustomizerlib/customizeboot.go

```diff
@@ -107,143 +93,228 @@ func replaceToken(inputGrubCfgContent string, oldToken string, newToken string)
+	end := initrdFilePathToken.Loc.End.Index
 
 	oldInitrdPath = inputGrubCfgContent[start:end]
-	outputGrubCfgContent = inputGrubCfgContent[:start] + initrdPath + inputGrubCfgContent[end:]
+	outputGrubCfgContent = inputGrubCfgContent[:start] + quotedInitrdPath + inputGrubCfgContent[end:]
 
 	return outputGrubCfgContent, oldInitrdPath, nil
 }
 
+// Appends kernel command-line args to the linux command within a grub config file.
 func appendKernelCommandLineArguments(inputGrubCfgContent string, extraCommandLine string) (outputGrubCfgContent string, err error) {
-	// Find the point where the new command line arguments should be inserted.
-	match := linuxCommandLineRegex.FindStringSubmatchIndex(inputGrubCfgContent)
-	if match == nil {
-		return "", fmt.Errorf("failed to find Linux kernel command line params in grub2 config file")
-	}
-	if len(match) <= linuxCommandLineRegexKernelOptsStart {
-		return "", fmt.Errorf("failed to find Linux kernel command line params in grub2 config file")
+	_, insertAtToken, err := getLinuxCommandLineArgs(inputGrubCfgContent)
+	if err != nil {
+		return "", err
 	}
 
-	start := match[linuxCommandLineRegexKernelOptsStart]
-
-	outputGrubCfgContent = inputGrubCfgContent[:start] + extraCommandLine + " " + inputGrubCfgContent[start:]
-
+	// Insert args at the end of the line.
+	insertPoint := insertAtToken.Loc.Start.Index
+	outputGrubCfgContent = inputGrubCfgContent[:insertPoint] + extraCommandLine + " " + inputGrubCfgContent[insertPoint:]
 	return outputGrubCfgContent, nil
 }
 
-// Tries to find the specified kernel CLI arg. Does not fail if the arg is not found.
+type grubConfigLinuxArg struct {
+	// The tokenizer token for the arg.
+	Token grub.Token
+	// The name of the argument.
+	Name string
+	// The value of the argument.
+	Value string
+	// If the argument's value has a variable expansion (e.g. $a).
+	ValueHasVarExpansion bool
+}
+
+// Finds the linux command within a grub config and returns a list of kernel command-line arguments.
 //
 // Returns:
-// - startIndex: The string index where the arg was found. Or -1 if not found.
-// - endIndex: The string index of the end of the arg. Or -1 if not found.
-// - value: The value of the arg. Or "" if not found.
-func findKernelCommandLineArgumentArg(grub2Config string, name string) (int, int, string, error) {
-	const (
-		argGroupIndex      = 2
-		argGroupStartIndex = argGroupIndex*2 + 0
-		argGroupEndIndex   = argGroupIndex*2 + 1
-
-		argValueGroupIndex      = 4
-		argValueGroupStartIndex = argValueGroupIndex*2 + 0
-		argValueGroupEndIndex   = argValueGroupIndex*2 + 1
-	)
-
-	escapedName := regexp.QuoteMeta(name)
-
-	// "linux ... <name>=<value> ..."
-	argumentRegexString := "(?m)[\\t ]+linux[\\t ](.*[\\t ])?(" + escapedName + "(=(\\S*)?)?)[\\t ]"
-	argumentRegex, err := regexp.Compile(argumentRegexString)
+// - args: A list of kernel command-line arguments.
+// - insertToken: A tokenizer token that represents an appropriate insert point for any new args.
```

Comment:
We probably need some clarification on what 'appropriate' means for the `insertToekn`. It sounds like it means right before `$kernelopts`?

Reply:
Added clarification to code comment.